### ISSUE TITLE
critical IO error on invalid SCSI CDB

### ISSUE THREAD

**Issue by jpsollie** [koverstreet/bcachefs#126]
```
[73647.492899] sd 12:0:1:0: [sdd] tag#679 UNKNOWN(0x2003) Result: hostbyte=0x00 driverbyte=0x08 cmd_age=0s
[73647.492902] sd 12:0:1:0: [sdd] tag#679 Sense Key : 0x5 [current] 
[73647.492903] sd 12:0:1:0: [sdd] tag#679 ASC=0x20 ASCQ=0x0 
[73647.492906] sd 12:0:1:0: [sdd] tag#679 CDB: opcode=0x35 35 00 00 00 00 00 00 00 00 00
[73647.492908] blk_update_request: critical target error, dev sdd, sector 0 op 0x2:(FLUSH) flags 0x0 phys_seg 0 prio class 0
[73647.492909] bcachefs (5044a9f4-27e4-4b73-bb89-4e104ab370eb): IO error on sdd for journal write
```

The cause here seems quite trivial: I disabled scsi_flush on aacraid controller because I wanted to avoid the RAID disks from spinning up every 5 min, and the on-board cache is backed up by a battery anyway.
The CDB specifies this command is reported as an illegal request.

Bcachefs doesn't seem to like it, while the answer from the SCSI is completely normal. re-enabling the SCSI flush command solves the problem

I know this is mostly a feature request, no bug, but just to report some exceptional cases.
I also have no idea what the "right" way to handle would be.

**Comment by koverstreet** [comment]
I think this needs to be fixed in the block layer. I'm not really a SCSI guy though - it seems might know more about scsi commands than I do. Can you explain "CBD" and what you mean by the scsi answer being completely normal?

**Comment by jpsollie** [comment]
CDB = command descriptor block.  a 10/16 byte sequence sent to a SCSI target to ask it to do something.
the seagate scsi reference manual may be useful here, just google "seagate scsi reference manual".
The host adapter (raid controller) sent a CDB beginning with 0x35 to the (virtual) raid target 12:0:1:0.  (host 12), (bus 1 is where the raid disks are stored), (device 0). 0x35 means CACHE FLUSH, whereby the virtual target replies with ASC 0x20, which means "illegal request", because I disabled all cache flush commands in the aacraid module.
The point here is that bcachefs should not worry about a cache flush being reported as an illegal request - in this particular case - which makes it complex: when should it worry, and when shouldn't it?
maybe I can find a CDB to query whether a cache flush is supported

**Comment by jpsollie** [comment]
The CDB query you need to look for is A3 0C 01 35 00 00 00 00 00 ff 00 00, which can be executed using sg_raw -r 256 /dev/sdx (CMD). If byte 2 has 101b or 011b flags, it should be fine. If the answer is 001b, it's unsupported. Unfortunately, still doesn't work for the aacraid arrays, they simply report queries for capacities are illegal requests ^^.
It's up to you what needs to happen here. Eventually, I could indeed ask the block layer guys for their opinion, but I don't really know which behaviour is supposed to happen...

**Comment by jpsollie** [comment]
fixed by block layer update in 5.13
